Description: Patches needed to enable LE architecture support for libvirt
Author: Chuck Short <zulcss@ubuntu.com>
Forwarded: Not-needed.
diff -Naurp libvirt-1.2.8.orig/docs/formatdomain.html.in libvirt-1.2.8/docs/formatdomain.html.in
--- libvirt-1.2.8.orig/docs/formatdomain.html.in	2014-08-29 08:22:31.000000000 -0400
+++ libvirt-1.2.8/docs/formatdomain.html.in	2014-11-27 08:54:34.097689984 -0500
@@ -1011,7 +1011,20 @@
           (such as CPUID level) that don't work. Until these issues are fixed,
           it's a good idea to avoid using <code>host-model</code> and use
           <code>custom</code> mode with just the CPU model from host
-          capabilities XML.</dd>
+          capabilities XML.
+          <span class="since">(Since 1.2.11)</span>. PowerISA allows
+          processors to run VMs in binary compatibility mode supporting an
+          older version of ISA.  Libvirt on PowerPC architecture uses the
+          <code>host-model</code> to signify a guest mode CPU running in
+          binary compatibility mode. Example:
+          When a user needs a power7 VM to run in compatibility mode
+          on a Power8 host, this can be described in XML as follows :
+<pre>
+  &lt;cpu mode='host-model'&gt;
+    &lt;model&gt;power7&lt;/model&gt;
+  &lt;/cpu&gt;
+  ...</pre>
+          </dd>
           <dt><code>host-passthrough</code></dt>
           <dd>With this mode, the CPU visible to the guest should be exactly
           the same as the host CPU even in the aspects that libvirt does not
diff -Naurp libvirt-1.2.8.orig/src/conf/cpu_conf.c libvirt-1.2.8/src/conf/cpu_conf.c
--- libvirt-1.2.8.orig/src/conf/cpu_conf.c	2014-07-17 02:47:28.000000000 -0400
+++ libvirt-1.2.8/src/conf/cpu_conf.c	2014-11-27 08:54:23.985689979 -0500
@@ -596,6 +596,7 @@ virCPUDefFormatBuf(virBufferPtr buf,
         return 0;
 
     formatModel = (def->mode == VIR_CPU_MODE_CUSTOM ||
+                   def->mode == VIR_CPU_MODE_HOST_MODEL ||
                    (flags & VIR_DOMAIN_XML_UPDATE_CPU));
     formatFallback = (def->type == VIR_CPU_TYPE_GUEST &&
                       (def->mode == VIR_CPU_MODE_HOST_MODEL ||
diff -Naurp libvirt-1.2.8.orig/src/cpu/cpu_map.xml libvirt-1.2.8/src/cpu/cpu_map.xml
--- libvirt-1.2.8.orig/src/cpu/cpu_map.xml	2014-07-17 02:47:28.000000000 -0400
+++ libvirt-1.2.8/src/cpu/cpu_map.xml	2014-11-27 08:54:29.193689981 -0500
@@ -627,5 +627,35 @@
       <pvr value='0x004b0100'/>
     </model>
 
+    <model name='power6'>
+      <vendor name='IBM'/>
+      <compat isa='2.05'/>
+      <pvr value='0x003e0000'/>
+    </model>
+
+    <model name='power7'>
+      <vendor name='IBM'/>
+      <compat isa='2.06'/>
+      <pvr value='0x003f0000'/>
+    </model>
+
+    <model name='power7+'>
+      <vendor name='IBM'/>
+      <compat isa='2.06B'/>
+      <pvr value='0x004a0000'/>
+    </model>
+
+    <model name='power8e'>
+      <vendor name='IBM'/>
+      <compat isa='2.07'/>
+      <pvr value='0x004b0000'/>
+    </model>
+
+    <model name='power8'>
+      <vendor name='IBM'/>
+      <compat isa='2.07'/>
+      <pvr value='0x004d0000'/>
+    </model>
+
   </arch>
 </cpus>
diff -Naurp libvirt-1.2.8.orig/src/cpu/cpu_powerpc.c libvirt-1.2.8/src/cpu/cpu_powerpc.c
--- libvirt-1.2.8.orig/src/cpu/cpu_powerpc.c	2014-06-25 21:49:55.000000000 -0400
+++ libvirt-1.2.8/src/cpu/cpu_powerpc.c	2014-11-27 08:54:29.197689981 -0500
@@ -99,6 +99,14 @@ ppcModelFindPVR(const struct ppc_map *ma
         model = model->next;
     }
 
+    /* PowerPC Processor Version Register is interpreted as follows :
+     * Higher order 16 bits : Power ISA generation.
+     * Lower order 16 bits : CPU chip version number.
+     * If the exact CPU isnt found, return the nearest matching CPU generation
+     */
+    if (pvr & 0x0000FFFFul)
+        return ppcModelFindPVR(map, (pvr & 0xFFFF0000ul));
+
     return NULL;
 }
 
@@ -562,8 +570,8 @@ ppcUpdate(virCPUDefPtr guest,
 static virCPUDefPtr
 ppcBaseline(virCPUDefPtr *cpus,
             unsigned int ncpus,
-            const char **models,
-            unsigned int nmodels,
+            const char **models ATTRIBUTE_UNUSED,
+            unsigned int nmodels ATTRIBUTE_UNUSED,
             unsigned int flags)
 {
     struct ppc_map *map = NULL;
@@ -583,13 +591,6 @@ ppcBaseline(virCPUDefPtr *cpus,
         goto error;
     }
 
-    if (!cpuModelIsAllowed(model->name, models, nmodels)) {
-        virReportError(VIR_ERR_CONFIG_UNSUPPORTED,
-                        _("CPU model %s is not supported by hypervisor"),
-                        model->name);
-        goto error;
-    }
-
     for (i = 0; i < ncpus; i++) {
         const struct ppc_vendor *vnd;
 
diff -Naurp libvirt-1.2.8.orig/src/qemu/qemu_capabilities.c libvirt-1.2.8/src/qemu/qemu_capabilities.c
--- libvirt-1.2.8.orig/src/qemu/qemu_capabilities.c	2014-08-29 08:22:31.000000000 -0400
+++ libvirt-1.2.8/src/qemu/qemu_capabilities.c	2014-11-27 08:54:18.777689976 -0500
@@ -631,7 +631,7 @@ virQEMUCapsProbeCPUModels(virQEMUCapsPtr
     if (qemuCaps->arch == VIR_ARCH_I686 ||
         qemuCaps->arch == VIR_ARCH_X86_64)
         parse = virQEMUCapsParseX86Models;
-    else if (qemuCaps->arch == VIR_ARCH_PPC64)
+    else if ARCH_IS_PPC64(qemuCaps->arch)
         parse = virQEMUCapsParsePPCModels;
     else {
         VIR_DEBUG("don't know how to parse %s CPU models",
@@ -1996,7 +1996,7 @@ bool virQEMUCapsHasPCIMultiBus(virQEMUCa
         return true;
 
     if (def->os.arch == VIR_ARCH_PPC ||
-        def->os.arch == VIR_ARCH_PPC64) {
+        ARCH_IS_PPC64(def->os.arch)) {
         /*
          * Usage of pci.0 naming:
          *
@@ -3567,7 +3567,7 @@ virQEMUCapsSupportsChardev(virDomainDefP
         !virQEMUCapsGet(qemuCaps, QEMU_CAPS_DEVICE))
         return false;
 
-    if ((def->os.arch == VIR_ARCH_PPC) || (def->os.arch == VIR_ARCH_PPC64)) {
+    if ((def->os.arch == VIR_ARCH_PPC) || ARCH_IS_PPC64(def->os.arch)) {
         /* only pseries need -device spapr-vty with -chardev */
         return (chr->deviceType == VIR_DOMAIN_CHR_DEVICE_TYPE_SERIAL &&
                 chr->info.type == VIR_DOMAIN_DEVICE_ADDRESS_TYPE_SPAPRVIO);
diff -Naurp libvirt-1.2.8.orig/src/qemu/qemu_command.c libvirt-1.2.8/src/qemu/qemu_command.c
--- libvirt-1.2.8.orig/src/qemu/qemu_command.c	2014-08-29 08:22:31.000000000 -0400
+++ libvirt-1.2.8/src/qemu/qemu_command.c	2014-11-27 08:54:23.989689979 -0500
@@ -703,7 +703,7 @@ qemuSetSCSIControllerModel(virDomainDefP
             return -1;
         }
     } else {
-        if ((def->os.arch == VIR_ARCH_PPC64) &&
+        if (ARCH_IS_PPC64(def->os.arch) &&
             STREQ(def->os.machine, "pseries")) {
             *model = VIR_DOMAIN_CONTROLLER_MODEL_SCSI_IBMVSCSI;
         } else if (virQEMUCapsGet(qemuCaps, QEMU_CAPS_SCSI_LSI)) {
@@ -1250,7 +1250,7 @@ int qemuDomainAssignSpaprVIOAddresses(vi
 
     for (i = 0; i < def->nserials; i++) {
         if (def->serials[i]->deviceType == VIR_DOMAIN_CHR_DEVICE_TYPE_SERIAL &&
-            (def->os.arch == VIR_ARCH_PPC64) &&
+            ARCH_IS_PPC64(def->os.arch) &&
             STREQ(def->os.machine, "pseries"))
             def->serials[i]->info.type = VIR_DOMAIN_DEVICE_ADDRESS_TYPE_SPAPRVIO;
         if (qemuAssignSpaprVIOAddress(def, &def->serials[i]->info,
@@ -1259,7 +1259,7 @@ int qemuDomainAssignSpaprVIOAddresses(vi
     }
 
     if (def->nvram) {
-        if (def->os.arch == VIR_ARCH_PPC64 &&
+        if (ARCH_IS_PPC64(def->os.arch) &&
             STREQ(def->os.machine, "pseries"))
             def->nvram->info.type = VIR_DOMAIN_DEVICE_ADDRESS_TYPE_SPAPRVIO;
         if (qemuAssignSpaprVIOAddress(def, &def->nvram->info,
@@ -4147,7 +4147,7 @@ qemuBuildUSBControllerDevStr(virDomainDe
     model = def->model;
 
     if (model == -1) {
-        if (domainDef->os.arch == VIR_ARCH_PPC64)
+        if ARCH_IS_PPC64(domainDef->os.arch)
             model = VIR_DOMAIN_CONTROLLER_MODEL_USB_PCI_OHCI;
         else
             model = VIR_DOMAIN_CONTROLLER_MODEL_USB_PIIX3_UHCI;
@@ -6052,64 +6052,52 @@ qemuBuildClockArgStr(virDomainClockDefPt
     return NULL;
 }
 
-
 static int
-qemuBuildCpuArgStr(virQEMUDriverPtr driver,
-                   const virDomainDef *def,
-                   const char *emulator,
-                   virQEMUCapsPtr qemuCaps,
-                   virArch hostarch,
-                   char **opt,
-                   bool *hasHwVirt,
-                   bool migrating)
+qemuBuildCpuModelArgStr(virQEMUDriverPtr driver,
+                        const virDomainDef *def,
+                        virBufferPtr buf,
+                        virQEMUCapsPtr qemuCaps,
+                        bool *hasHwVirt,
+                        bool migrating)
 {
+    int ret = -1;
+    size_t i;
     virCPUDefPtr host = NULL;
     virCPUDefPtr guest = NULL;
     virCPUDefPtr cpu = NULL;
     size_t ncpus = 0;
     char **cpus = NULL;
-    const char *default_model;
     virCPUDataPtr data = NULL;
-    bool have_cpu = false;
     char *compare_msg = NULL;
-    int ret = -1;
-    virBuffer buf = VIR_BUFFER_INITIALIZER;
-    size_t i;
+    virCPUCompareResult cmp;
+    const char *preferred;
     virCapsPtr caps = NULL;
-
-    *hasHwVirt = false;
+    bool compareAgainstHost = (def->virtType == VIR_DOMAIN_VIRT_KVM ||
+        def->cpu->mode != VIR_CPU_MODE_CUSTOM);
 
     if (!(caps = virQEMUDriverGetCapabilities(driver, false)))
         goto cleanup;
 
     host = caps->host.cpu;
 
-    if (def->os.arch == VIR_ARCH_I686)
-        default_model = "qemu32";
-    else
-        default_model = "qemu64";
-
-    if (def->cpu &&
-        (def->cpu->mode != VIR_CPU_MODE_CUSTOM || def->cpu->model)) {
-        virCPUCompareResult cmp;
-        const char *preferred;
-
-        if (!host ||
-            !host->model ||
-            (ncpus = virQEMUCapsGetCPUDefinitions(qemuCaps, &cpus)) == 0) {
-            virReportError(VIR_ERR_CONFIG_UNSUPPORTED, "%s",
-                           _("CPU specification not supported by hypervisor"));
-            goto cleanup;
-        }
+    if (!host ||
+        !host->model ||
+        (ncpus = virQEMUCapsGetCPUDefinitions(qemuCaps, &cpus)) == 0) {
+        virReportError(VIR_ERR_CONFIG_UNSUPPORTED, "%s",
+                       _("CPU specification not supported by hypervisor"));
+        goto cleanup;
+    }
 
-        if (!(cpu = virCPUDefCopy(def->cpu)))
-            goto cleanup;
+    if (!(cpu = virCPUDefCopy(def->cpu)))
+        goto cleanup;
 
-        if (cpu->mode != VIR_CPU_MODE_CUSTOM &&
-            !migrating &&
-            cpuUpdate(cpu, host) < 0)
-            goto cleanup;
+    if (cpu->mode != VIR_CPU_MODE_CUSTOM &&
+        !migrating &&
+        cpuUpdate(cpu, host) < 0)
+        goto cleanup;
 
+    /* For non-KVM, CPU features are emulated, so host compat doesn't matter */
+    if (compareAgainstHost) {
         cmp = cpuGuestData(host, cpu, &data, &compare_msg);
         switch (cmp) {
         case VIR_CPU_COMPARE_INCOMPATIBLE:
@@ -6128,39 +6116,49 @@ qemuBuildCpuArgStr(virQEMUDriverPtr driv
         default:
             break;
         }
+    }
 
-        /* Only 'svm' requires --enable-nesting. The nested
-         * 'vmx' patches now simply hook off the CPU features
-         */
-        if (def->os.arch == VIR_ARCH_X86_64 ||
-            def->os.arch == VIR_ARCH_I686) {
-            int hasSVM = cpuHasFeature(data, "svm");
-            if (hasSVM < 0)
-                goto cleanup;
-            *hasHwVirt = hasSVM > 0 ? true : false;
+    /* Only 'svm' requires --enable-nesting. The nested
+     * 'vmx' patches now simply hook off the CPU features
+     */
+    if ((def->os.arch == VIR_ARCH_X86_64 || def->os.arch == VIR_ARCH_I686) &&
+         compareAgainstHost) {
+        int hasSVM = cpuHasFeature(data, "svm");
+        if (hasSVM < 0)
+            goto cleanup;
+        *hasHwVirt = hasSVM > 0 ? true : false;
+    }
+
+    if ((cpu->mode == VIR_CPU_MODE_HOST_PASSTHROUGH) ||
+        ((cpu->mode == VIR_CPU_MODE_HOST_MODEL) &&
+          ARCH_IS_PPC64(def->os.arch))) {
+        const char *mode = virCPUModeTypeToString(cpu->mode);
+        if (!virQEMUCapsGet(qemuCaps, QEMU_CAPS_CPU_HOST)) {
+            virReportError(VIR_ERR_CONFIG_UNSUPPORTED,
+                           _("CPU mode '%s' is not supported by QEMU"
+                             " binary"), mode);
+            goto cleanup;
         }
+        if (def->virtType != VIR_DOMAIN_VIRT_KVM) {
+            virReportError(VIR_ERR_CONFIG_UNSUPPORTED,
+                           _("CPU mode '%s' is only supported with kvm"),
+                           mode);
+            goto cleanup;
+        }
+        virBufferAddLit(buf, "host");
 
-        if (cpu->mode == VIR_CPU_MODE_HOST_PASSTHROUGH) {
-            const char *mode = virCPUModeTypeToString(cpu->mode);
-            if (!virQEMUCapsGet(qemuCaps, QEMU_CAPS_CPU_HOST)) {
-                virReportError(VIR_ERR_CONFIG_UNSUPPORTED,
-                               _("CPU mode '%s' is not supported by QEMU"
-                                 " binary"), mode);
-                goto cleanup;
-            }
-            if (def->virtType != VIR_DOMAIN_VIRT_KVM) {
-                virReportError(VIR_ERR_CONFIG_UNSUPPORTED,
-                               _("CPU mode '%s' is only supported with kvm"),
-                               mode);
-                goto cleanup;
-            }
-            virBufferAddLit(&buf, "host");
-        } else {
-            if (VIR_ALLOC(guest) < 0)
-                goto cleanup;
-            if (VIR_STRDUP(guest->vendor_id, cpu->vendor_id) < 0)
-                goto cleanup;
+        if (ARCH_IS_PPC64(def->os.arch) &&
+            cpu->mode == VIR_CPU_MODE_HOST_MODEL) {
+            virBufferAsprintf(buf, ",compat=%s", def->cpu->model);
+        }
 
+    } else {
+        if (VIR_ALLOC(guest) < 0)
+            goto cleanup;
+        if (VIR_STRDUP(guest->vendor_id, cpu->vendor_id) < 0)
+            goto cleanup;
+
+        if (compareAgainstHost) {
             guest->arch = host->arch;
             if (cpu->match == VIR_CPU_MATCH_MINIMUM)
                 preferred = host->model;
@@ -6169,22 +6167,67 @@ qemuBuildCpuArgStr(virQEMUDriverPtr driv
 
             guest->type = VIR_CPU_TYPE_GUEST;
             guest->fallback = cpu->fallback;
-            if (cpuDecode(guest, data, (const char **)cpus, ncpus, preferred) < 0)
+            if (cpuDecode(guest, data,
+                          (const char **)cpus, ncpus, preferred) < 0)
+                goto cleanup;
+        } else {
+            guest->arch = def->os.arch;
+            if (VIR_STRDUP(guest->model, cpu->model) < 0)
                 goto cleanup;
+        }
 
-            virBufferAdd(&buf, guest->model, -1);
-            if (guest->vendor_id)
-                virBufferAsprintf(&buf, ",vendor=%s", guest->vendor_id);
-            for (i = 0; i < guest->nfeatures; i++) {
-                char sign;
-                if (guest->features[i].policy == VIR_CPU_FEATURE_DISABLE)
-                    sign = '-';
-                else
-                    sign = '+';
+        virBufferAdd(buf, guest->model, -1);
+        if (guest->vendor_id)
+            virBufferAsprintf(buf, ",vendor=%s", guest->vendor_id);
+        for (i = 0; i < guest->nfeatures; i++) {
+            char sign;
+            if (guest->features[i].policy == VIR_CPU_FEATURE_DISABLE)
+                sign = '-';
+            else
+                sign = '+';
 
-                virBufferAsprintf(&buf, ",%c%s", sign, guest->features[i].name);
-            }
+            virBufferAsprintf(buf, ",%c%s", sign, guest->features[i].name);
         }
+    }
+
+    ret = 0;
+ cleanup:
+    virObjectUnref(caps);
+    VIR_FREE(compare_msg);
+    cpuDataFree(data);
+    virCPUDefFree(guest);
+    virCPUDefFree(cpu);
+    return ret;
+}
+
+static int
+qemuBuildCpuArgStr(virQEMUDriverPtr driver,
+                   const virDomainDef *def,
+                   const char *emulator,
+                   virQEMUCapsPtr qemuCaps,
+                   virArch hostarch,
+                   char **opt,
+                   bool *hasHwVirt,
+                   bool migrating)
+{
+    const char *default_model;
+    bool have_cpu = false;
+    int ret = -1;
+    virBuffer buf = VIR_BUFFER_INITIALIZER;
+    size_t i;
+
+    *hasHwVirt = false;
+
+    if (def->os.arch == VIR_ARCH_I686)
+        default_model = "qemu32";
+    else
+        default_model = "qemu64";
+
+    if (def->cpu &&
+        (def->cpu->mode != VIR_CPU_MODE_CUSTOM || def->cpu->model)) {
+        if (qemuBuildCpuModelArgStr(driver, def, &buf, qemuCaps,
+                                    hasHwVirt, migrating) < 0)
+            goto cleanup;
         have_cpu = true;
     } else {
         /*
@@ -6309,11 +6352,6 @@ qemuBuildCpuArgStr(virQEMUDriverPtr driv
     ret = 0;
 
  cleanup:
-    VIR_FREE(compare_msg);
-    cpuDataFree(data);
-    virCPUDefFree(guest);
-    virCPUDefFree(cpu);
-    virObjectUnref(caps);
     return ret;
 }
 
@@ -8176,7 +8214,7 @@ qemuBuildCommandLine(virConnectPtr conn,
                            !qemuDomainMachineIsQ35(def) &&
                            (!virQEMUCapsGet(qemuCaps, QEMU_CAPS_PIIX3_USB_UHCI) ||
                             (!virQEMUCapsGet(qemuCaps, QEMU_CAPS_PCI_OHCI) &&
-                             def->os.arch == VIR_ARCH_PPC64))) {
+                             ARCH_IS_PPC64(def->os.arch)))) {
                     if (usblegacy) {
                         virReportError(VIR_ERR_CONFIG_UNSUPPORTED, "%s",
                                        _("Multiple legacy USB controllers are "
@@ -9366,7 +9404,7 @@ qemuBuildCommandLine(virConnectPtr conn,
     }
 
     if (def->nvram) {
-        if (def->os.arch == VIR_ARCH_PPC64 &&
+        if (ARCH_IS_PPC64(def->os.arch) &&
             STREQ(def->os.machine, "pseries")) {
             if (!virQEMUCapsGet(qemuCaps, QEMU_CAPS_DEVICE_NVRAM)) {
                 virReportError(VIR_ERR_CONFIG_UNSUPPORTED, "%s",
@@ -9478,7 +9516,7 @@ qemuBuildSerialChrDeviceStr(char **devic
 {
     virBuffer cmd = VIR_BUFFER_INITIALIZER;
 
-    if ((arch == VIR_ARCH_PPC64) && STREQ(machine, "pseries")) {
+    if (ARCH_IS_PPC64(arch) && STREQ(machine, "pseries")) {
         if (serial->deviceType == VIR_DOMAIN_CHR_DEVICE_TYPE_SERIAL &&
             serial->info.type == VIR_DOMAIN_DEVICE_ADDRESS_TYPE_SPAPRVIO) {
             virBufferAsprintf(&cmd, "spapr-vty,chardev=char%s",
@@ -9900,7 +9938,7 @@ qemuParseCommandLineDisk(virDomainXMLOpt
     if (VIR_ALLOC(def->src) < 0)
         goto error;
 
-    if (((dom->os.arch == VIR_ARCH_PPC64) &&
+    if ((ARCH_IS_PPC64(dom->os.arch) &&
         dom->os.machine && STREQ(dom->os.machine, "pseries")))
         def->bus = VIR_DOMAIN_DISK_BUS_SCSI;
     else
@@ -11233,7 +11271,7 @@ qemuParseCommandLine(virCapsPtr qemuCaps
                 disk->src->type = VIR_STORAGE_TYPE_FILE;
             if (STREQ(arg, "-cdrom")) {
                 disk->device = VIR_DOMAIN_DISK_DEVICE_CDROM;
-                if (((def->os.arch == VIR_ARCH_PPC64) &&
+                if ((ARCH_IS_PPC64(def->os.arch) &&
                     def->os.machine && STREQ(def->os.machine, "pseries")))
                     disk->bus = VIR_DOMAIN_DISK_BUS_SCSI;
                 if (VIR_STRDUP(disk->dst, "hdc") < 0)
diff -Naurp libvirt-1.2.8.orig/src/qemu/qemu_domain.c libvirt-1.2.8/src/qemu/qemu_domain.c
--- libvirt-1.2.8.orig/src/qemu/qemu_domain.c	2014-08-18 22:47:32.000000000 -0400
+++ libvirt-1.2.8/src/qemu/qemu_domain.c	2014-11-27 08:54:18.785689976 -0500
@@ -754,6 +754,7 @@ qemuDomainDefPostParse(virDomainDefPtr d
        break;
 
     case VIR_ARCH_PPC64:
+    case VIR_ARCH_PPC64LE:
         addPCIRoot = true;
         addDefaultUSBKBD = true;
         addDefaultUSBMouse = true;
diff -Naurp libvirt-1.2.8.orig/src/util/virarch.h libvirt-1.2.8/src/util/virarch.h
--- libvirt-1.2.8.orig/src/util/virarch.h	2014-06-25 07:25:52.000000000 -0400
+++ libvirt-1.2.8/src/util/virarch.h	2014-11-27 08:54:18.785689976 -0500
@@ -79,6 +79,9 @@ typedef enum {
                              (arch) == VIR_ARCH_PPC64LE ||\
                              (arch) == VIR_ARCH_PPCEMB)
 
+# define ARCH_IS_PPC64(arch)  ((arch) == VIR_ARCH_PPC64 ||\
+                             (arch) == VIR_ARCH_PPC64LE)
+
 # define ARCH_IS_ARM(arch)  ((arch) == VIR_ARCH_ARMV6L ||\
                              (arch) == VIR_ARCH_ARMV7L ||\
                              (arch) == VIR_ARCH_ARMV7B ||\
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-aarch64-virt-default-nic.args libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-aarch64-virt-default-nic.args
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-aarch64-virt-default-nic.args	2014-02-21 06:24:40.000000000 -0500
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-aarch64-virt-default-nic.args	2014-11-27 08:54:12.649689973 -0500
@@ -1,5 +1,6 @@
 LC_ALL=C PATH=/bin HOME=/home/test USER=test LOGNAME=test QEMU_AUDIO_DRV=none \
-/usr/bin/qemu-system-aarch64 -S -M virt -m 1024 -smp 1 -nographic \
+/usr/bin/qemu-system-aarch64 -S -M virt -cpu cortex-a53 \
+-m 1024 -smp 1 -nographic \
 -nodefconfig -nodefaults -monitor unix:/tmp/test-monitor,server,nowait \
 -boot c -kernel /aarch64.kernel -initrd /aarch64.initrd -append console=ttyAMA0 \
 -usb -device virtio-net-device,vlan=0,id=net0,mac=52:54:00:09:a4:37 \
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-aarch64-virt-default-nic.xml libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-aarch64-virt-default-nic.xml
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-aarch64-virt-default-nic.xml	2014-02-21 06:24:40.000000000 -0500
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-aarch64-virt-default-nic.xml	2014-11-27 08:54:12.649689973 -0500
@@ -7,6 +7,9 @@
   <features>
     <acpi/>
   </features>
+  <cpu match='exact'>
+    <model>cortex-a53</model>
+  </cpu>
   <os>
     <type arch="aarch64" machine="virt">hvm</type>
     <kernel>/aarch64.kernel</kernel>
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-aarch64-virt-virtio.args libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-aarch64-virt-virtio.args
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-aarch64-virt-virtio.args	2014-01-07 22:12:37.000000000 -0500
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-aarch64-virt-virtio.args	2014-11-27 08:54:12.649689973 -0500
@@ -1,5 +1,6 @@
 LC_ALL=C PATH=/bin HOME=/home/test USER=test LOGNAME=test QEMU_AUDIO_DRV=none \
-/usr/bin/qemu-system-aarch64 -S -M virt -m 1024 -smp 1 -nographic \
+/usr/bin/qemu-system-aarch64 -S -M virt -cpu cortex-a53 \
+-m 1024 -smp 1 -nographic \
 -nodefconfig -nodefaults -monitor unix:/tmp/test-monitor,server,nowait \
 -boot c -kernel /aarch64.kernel -initrd /aarch64.initrd -append \
 'earlyprintk console=ttyAMA0,115200n8 rw root=/dev/vda rootwait' \
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-aarch64-virt-virtio.xml libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-aarch64-virt-virtio.xml
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-aarch64-virt-virtio.xml	2014-01-07 22:12:37.000000000 -0500
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-aarch64-virt-virtio.xml	2014-11-27 08:54:12.649689973 -0500
@@ -16,6 +16,9 @@
     <apic/>
     <pae/>
   </features>
+  <cpu match='exact'>
+    <model>cortex-a53</model>
+  </cpu>
   <clock offset="utc"/>
   <on_poweroff>destroy</on_poweroff>
   <on_reboot>restart</on_reboot>
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact1.args libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact1.args
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact1.args	2013-09-12 05:18:15.000000000 -0400
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact1.args	2014-11-27 08:54:12.649689973 -0500
@@ -1,5 +1,5 @@
 LC_ALL=C PATH=/bin HOME=/home/test USER=test LOGNAME=test QEMU_AUDIO_DRV=none \
-/usr/bin/qemu -S -M pc \
+/usr/bin/qemu-kvm -S -M pc \
 -cpu qemu64,-svm,-lm,-nx,-syscall,-clflush,-pse36,-mca -m 214 -smp 6 \
 -nographic -monitor unix:/tmp/test-monitor,server,nowait -no-acpi -boot n -usb -net \
 none -serial none -parallel none
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact1.xml libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact1.xml
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact1.xml	2012-09-27 23:44:10.000000000 -0400
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact1.xml	2014-11-27 08:54:12.649689973 -0500
@@ -1,4 +1,4 @@
-<domain type='qemu'>
+<domain type='kvm'>
   <name>QEMUGuest1</name>
   <uuid>c7a5fdbd-edaf-9455-926a-d65c16db1809</uuid>
   <memory unit='KiB'>219100</memory>
@@ -23,6 +23,6 @@
   <on_reboot>restart</on_reboot>
   <on_crash>destroy</on_crash>
   <devices>
-      <emulator>/usr/bin/qemu</emulator>
+      <emulator>/usr/bin/qemu-kvm</emulator>
   </devices>
 </domain>
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact2.args libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact2.args
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact2.args	2013-09-12 05:18:15.000000000 -0400
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact2.args	2014-11-27 08:54:12.649689973 -0500
@@ -1,5 +1,5 @@
 LC_ALL=C PATH=/bin HOME=/home/test USER=test LOGNAME=test QEMU_AUDIO_DRV=none \
-/usr/bin/qemu -S -M pc \
+/usr/bin/qemu-kvm -S -M pc \
 -cpu core2duo,+lahf_lm,+3dnowext,+xtpr,+ds_cpl,+tm,+ht,+ds,-nx -m 214 -smp 6 \
 -nographic -monitor unix:/tmp/test-monitor,server,nowait -no-acpi -boot n -usb -net \
 none -serial none -parallel none
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact2-nofallback.args libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact2-nofallback.args
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact2-nofallback.args	2013-09-12 05:18:15.000000000 -0400
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact2-nofallback.args	2014-11-27 08:54:12.649689973 -0500
@@ -1,5 +1,5 @@
 LC_ALL=C PATH=/bin HOME=/home/test USER=test LOGNAME=test QEMU_AUDIO_DRV=none \
-/usr/bin/qemu -S -M pc \
+/usr/bin/qemu-kvm -S -M pc \
 -cpu core2duo,+lahf_lm,+3dnowext,+xtpr,+ds_cpl,+tm,+ht,+ds,-nx -m 214 -smp 6 \
 -nographic -monitor unix:/tmp/test-monitor,server,nowait -no-acpi -boot n -usb -net \
 none -serial none -parallel none
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact2-nofallback.xml libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact2-nofallback.xml
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact2-nofallback.xml	2012-09-27 23:44:10.000000000 -0400
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact2-nofallback.xml	2014-11-27 08:54:12.649689973 -0500
@@ -1,4 +1,4 @@
-<domain type='qemu'>
+<domain type='kvm'>
   <name>QEMUGuest1</name>
   <uuid>c7a5fdbd-edaf-9455-926a-d65c16db1809</uuid>
   <memory unit='KiB'>219100</memory>
@@ -30,6 +30,6 @@
   <on_reboot>restart</on_reboot>
   <on_crash>destroy</on_crash>
   <devices>
-      <emulator>/usr/bin/qemu</emulator>
+      <emulator>/usr/bin/qemu-kvm</emulator>
   </devices>
 </domain>
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact2.xml libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact2.xml
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact2.xml	2012-09-27 23:44:10.000000000 -0400
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-exact2.xml	2014-11-27 08:54:12.649689973 -0500
@@ -1,4 +1,4 @@
-<domain type='qemu'>
+<domain type='kvm'>
   <name>QEMUGuest1</name>
   <uuid>c7a5fdbd-edaf-9455-926a-d65c16db1809</uuid>
   <memory unit='KiB'>219100</memory>
@@ -30,6 +30,6 @@
   <on_reboot>restart</on_reboot>
   <on_crash>destroy</on_crash>
   <devices>
-      <emulator>/usr/bin/qemu</emulator>
+      <emulator>/usr/bin/qemu-kvm</emulator>
   </devices>
 </domain>
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-fallback.args libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-fallback.args
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-fallback.args	2013-09-12 05:18:15.000000000 -0400
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-fallback.args	2014-11-27 08:54:12.653689973 -0500
@@ -3,7 +3,7 @@ PATH=/bin \
 HOME=/home/test \
 USER=test \
 LOGNAME=test QEMU_AUDIO_DRV=none \
-/usr/bin/qemu \
+/usr/bin/qemu-kvm \
 -S \
 -M pc \
 -cpu Penryn,-sse4.1 \
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-fallback.xml libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-fallback.xml
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-fallback.xml	2012-09-27 23:44:10.000000000 -0400
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-fallback.xml	2014-11-27 08:54:12.653689973 -0500
@@ -1,4 +1,4 @@
-<domain type='qemu'>
+<domain type='kvm'>
   <name>QEMUGuest1</name>
   <uuid>c7a5fdbd-edaf-9455-926a-d65c16db1809</uuid>
   <memory unit='KiB'>219100</memory>
@@ -20,6 +20,6 @@
   <on_reboot>restart</on_reboot>
   <on_crash>destroy</on_crash>
   <devices>
-      <emulator>/usr/bin/qemu</emulator>
+      <emulator>/usr/bin/qemu-kvm</emulator>
   </devices>
 </domain>
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-minimum1.args libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-minimum1.args
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-minimum1.args	2013-09-12 05:18:15.000000000 -0400
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-minimum1.args	2014-11-27 08:54:12.653689973 -0500
@@ -1,5 +1,5 @@
 LC_ALL=C PATH=/bin HOME=/home/test USER=test LOGNAME=test QEMU_AUDIO_DRV=none \
-/usr/bin/qemu -S -M pc \
+/usr/bin/qemu-kvm -S -M pc \
 -cpu core2duo,+lahf_lm,+xtpr,+cx16,+tm2,+est,+vmx,+ds_cpl,+pbe,+tm,+ht,+ss,\
 +acpi,+ds -m 214 -smp 6 -nographic -monitor unix:/tmp/test-monitor,server,\
 nowait -no-acpi -boot n -usb -net none -serial none -parallel none
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-minimum1.xml libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-minimum1.xml
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-minimum1.xml	2012-09-27 23:44:10.000000000 -0400
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-minimum1.xml	2014-11-27 08:54:12.653689973 -0500
@@ -1,4 +1,4 @@
-<domain type='qemu'>
+<domain type='kvm'>
   <name>QEMUGuest1</name>
   <uuid>c7a5fdbd-edaf-9455-926a-d65c16db1809</uuid>
   <memory unit='KiB'>219100</memory>
@@ -16,6 +16,6 @@
   <on_reboot>restart</on_reboot>
   <on_crash>destroy</on_crash>
   <devices>
-      <emulator>/usr/bin/qemu</emulator>
+      <emulator>/usr/bin/qemu-kvm</emulator>
   </devices>
 </domain>
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-minimum2.args libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-minimum2.args
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-minimum2.args	2013-09-12 05:18:15.000000000 -0400
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-minimum2.args	2014-11-27 08:54:12.653689973 -0500
@@ -1,5 +1,5 @@
 LC_ALL=C PATH=/bin HOME=/home/test USER=test LOGNAME=test QEMU_AUDIO_DRV=none \
-/usr/bin/qemu -S -M pc \
+/usr/bin/qemu-kvm -S -M pc \
 -cpu core2duo,+lahf_lm,+xtpr,+cx16,+tm2,+est,+vmx,+ds_cpl,+pbe,+tm,+ht,+ss,\
 +acpi,+ds,-lm,-nx,-syscall -m 214 -smp 6 -nographic -monitor \
 unix:/tmp/test-monitor,server,nowait -no-acpi -boot n -usb -net none -serial none \
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-minimum2.xml libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-minimum2.xml
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-minimum2.xml	2012-09-27 23:44:10.000000000 -0400
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-minimum2.xml	2014-11-27 08:54:12.653689973 -0500
@@ -1,4 +1,4 @@
-<domain type='qemu'>
+<domain type='kvm'>
   <name>QEMUGuest1</name>
   <uuid>c7a5fdbd-edaf-9455-926a-d65c16db1809</uuid>
   <memory unit='KiB'>219100</memory>
@@ -20,6 +20,6 @@
   <on_reboot>restart</on_reboot>
   <on_crash>destroy</on_crash>
   <devices>
-      <emulator>/usr/bin/qemu</emulator>
+      <emulator>/usr/bin/qemu-kvm</emulator>
   </devices>
 </domain>
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-nofallback.xml libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-nofallback.xml
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-nofallback.xml	2012-09-27 23:44:10.000000000 -0400
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-nofallback.xml	2014-11-27 08:54:12.653689973 -0500
@@ -1,4 +1,4 @@
-<domain type='qemu'>
+<domain type='kvm'>
   <name>QEMUGuest1</name>
   <uuid>c7a5fdbd-edaf-9455-926a-d65c16db1809</uuid>
   <memory unit='KiB'>219100</memory>
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-strict1.args libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-strict1.args
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-strict1.args	2013-09-12 05:18:15.000000000 -0400
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-strict1.args	2014-11-27 08:54:12.653689973 -0500
@@ -1,5 +1,5 @@
 LC_ALL=C PATH=/bin HOME=/home/test USER=test LOGNAME=test QEMU_AUDIO_DRV=none \
-/usr/bin/qemu -S -M pc \
+/usr/bin/qemu-kvm -S -M pc \
 -cpu core2duo,+lahf_lm,+3dnowext,+xtpr,+est,+vmx,+ds_cpl,+tm,+ht,+acpi,+ds,-nx \
 -m 214 -smp 6 -nographic -monitor unix:/tmp/test-monitor,server,nowait \
 -no-acpi -boot n -usb -net none -serial none -parallel none
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-strict1.xml libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-strict1.xml
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-cpu-strict1.xml	2012-09-27 23:44:10.000000000 -0400
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-cpu-strict1.xml	2014-11-27 08:54:12.653689973 -0500
@@ -1,4 +1,4 @@
-<domain type='qemu'>
+<domain type='kvm'>
   <name>QEMUGuest1</name>
   <uuid>c7a5fdbd-edaf-9455-926a-d65c16db1809</uuid>
   <memory unit='KiB'>219100</memory>
@@ -33,6 +33,6 @@
   <on_reboot>restart</on_reboot>
   <on_crash>destroy</on_crash>
   <devices>
-      <emulator>/usr/bin/qemu</emulator>
+      <emulator>/usr/bin/qemu-kvm</emulator>
   </devices>
 </domain>
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-graphics-spice-timeout.args libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-graphics-spice-timeout.args
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-graphics-spice-timeout.args	2013-07-23 20:23:07.000000000 -0400
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-graphics-spice-timeout.args	2014-11-27 08:54:12.653689973 -0500
@@ -1,5 +1,5 @@
 LC_ALL=C PATH=/bin HOME=/home/test USER=test LOGNAME=test QEMU_AUDIO_DRV=spice \
-/usr/bin/qemu -S -M pc -cpu core2duo,+lahf_lm,+xtpr,+cx16,+tm2,\
+/usr/bin/qemu-kvm -S -M pc -cpu core2duo,+lahf_lm,+xtpr,+cx16,+tm2,\
 +est,+vmx,+ds_cpl,+pbe,+tm,+ht,+ss,+acpi,+ds \
 -m 1024 -smp 2 -nodefaults -monitor unix:/tmp/test-monitor,server,nowait \
 -boot dc -device virtio-serial-pci,id=virtio-serial0,bus=pci.0,addr=0x6 \
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-graphics-spice-timeout.xml libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-graphics-spice-timeout.xml
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-graphics-spice-timeout.xml	2014-02-21 06:24:40.000000000 -0500
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-graphics-spice-timeout.xml	2014-11-27 08:54:12.653689973 -0500
@@ -1,4 +1,4 @@
-<domain type='qemu'>
+<domain type='kvm'>
   <name>f14</name>
   <uuid>553effab-b5e1-2d80-dfe3-da4344826c43</uuid>
   <memory unit='KiB'>1048576</memory>
@@ -38,7 +38,7 @@
   <on_reboot>restart</on_reboot>
   <on_crash>restart</on_crash>
   <devices>
-    <emulator>/usr/bin/qemu</emulator>
+    <emulator>/usr/bin/qemu-kvm</emulator>
     <disk type='file' device='disk'>
       <driver name='qemu' type='qcow2'/>
       <source file='/var/lib/libvirt/images/f14.img'/>
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-pseries-cpu-compat.args libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-pseries-cpu-compat.args
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-pseries-cpu-compat.args	1969-12-31 19:00:00.000000000 -0500
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-pseries-cpu-compat.args	2014-11-27 08:54:38.889689986 -0500
@@ -0,0 +1,8 @@
+LC_ALL=C PATH=/bin HOME=/home/test USER=test LOGNAME=test \
+QEMU_AUDIO_DRV=none /usr/bin/qemu-system-ppc64 -S -M pseries \
+-cpu host,compat=power7 \
+-m 214 -smp 4 -nographic -nodefconfig -nodefaults \
+-chardev socket,id=charmonitor,path=/tmp/test-monitor,server,nowait \
+-mon chardev=charmonitor,id=monitor,mode=readline -no-acpi -boot c -usb \
+-chardev pty,id=charserial0 \
+-device spapr-vty,chardev=charserial0,reg=0x30000000
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-pseries-cpu-compat.xml libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-pseries-cpu-compat.xml
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-pseries-cpu-compat.xml	1969-12-31 19:00:00.000000000 -0500
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-pseries-cpu-compat.xml	2014-11-27 08:54:38.889689986 -0500
@@ -0,0 +1,20 @@
+<domain type='kvm'>
+  <name>QEMUGuest1</name>
+  <memory unit='KiB'>219100</memory>
+  <currentMemory unit='KiB'>219100</currentMemory>
+  <vcpu placement='static'>4</vcpu>
+  <os>
+    <type arch='ppc64' machine='pseries'>hvm</type>
+  </os>
+  <cpu mode='host-model'>
+    <model>power7</model>
+  </cpu>
+  <clock offset='utc'/>
+  <devices>
+      <emulator>/usr/bin/qemu-system-ppc64</emulator>
+      <console type='pty'>
+        <address type="spapr-vio"/>
+      </console>
+      <memballoon model="none"/>
+  </devices>
+</domain>
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-pseries-cpu-exact.args libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-pseries-cpu-exact.args
--- libvirt-1.2.8.orig/tests/qemuxml2argvdata/qemuxml2argv-pseries-cpu-exact.args	2013-09-12 05:18:15.000000000 -0400
+++ libvirt-1.2.8/tests/qemuxml2argvdata/qemuxml2argv-pseries-cpu-exact.args	2014-11-27 08:54:12.653689973 -0500
@@ -1,6 +1,6 @@
 LC_ALL=C PATH=/bin HOME=/home/test USER=test LOGNAME=test \
-/usr/bin/qemu-system-ppc64 -S -M pseries -cpu POWER7_v2.3 -m 512 -smp 1 -nographic \
--nodefconfig -nodefaults \
+QEMU_AUDIO_DRV=none /usr/bin/qemu-system-ppc64 -S -M pseries -cpu POWER7_v2.3 \
+-m 512 -smp 1 -nographic -nodefconfig -nodefaults \
 -chardev socket,id=charmonitor,path=/tmp/test-monitor,server,nowait \
 -mon chardev=charmonitor,id=monitor,mode=readline -no-acpi -boot c -usb \
 -chardev pty,id=charserial0 \
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2argvtest.c libvirt-1.2.8/tests/qemuxml2argvtest.c
--- libvirt-1.2.8.orig/tests/qemuxml2argvtest.c	2014-08-29 08:22:31.000000000 -0400
+++ libvirt-1.2.8/tests/qemuxml2argvtest.c	2014-11-27 08:54:38.889689986 -0500
@@ -920,7 +920,7 @@ mymain(void)
             QEMU_CAPS_DEVICE, QEMU_CAPS_SPICE,
             QEMU_CAPS_DEVICE_QXL);
     DO_TEST("graphics-spice-timeout",
-            QEMU_CAPS_DRIVE,
+            QEMU_CAPS_KVM, QEMU_CAPS_DRIVE,
             QEMU_CAPS_VGA, QEMU_CAPS_VGA_QXL,
             QEMU_CAPS_DEVICE, QEMU_CAPS_SPICE,
             QEMU_CAPS_DEVICE_QXL_VGA);
@@ -1192,14 +1192,14 @@ mymain(void)
     DO_TEST("cpu-topology1", QEMU_CAPS_SMP_TOPOLOGY);
     DO_TEST("cpu-topology2", QEMU_CAPS_SMP_TOPOLOGY);
     DO_TEST("cpu-topology3", NONE);
-    DO_TEST("cpu-minimum1", NONE);
-    DO_TEST("cpu-minimum2", NONE);
-    DO_TEST("cpu-exact1", NONE);
-    DO_TEST("cpu-exact2", NONE);
-    DO_TEST("cpu-exact2-nofallback", NONE);
-    DO_TEST("cpu-fallback", NONE);
-    DO_TEST_FAILURE("cpu-nofallback", NONE);
-    DO_TEST("cpu-strict1", NONE);
+    DO_TEST("cpu-minimum1", QEMU_CAPS_KVM);
+    DO_TEST("cpu-minimum2", QEMU_CAPS_KVM);
+    DO_TEST("cpu-exact1", QEMU_CAPS_KVM);
+    DO_TEST("cpu-exact2", QEMU_CAPS_KVM);
+    DO_TEST("cpu-exact2-nofallback", QEMU_CAPS_KVM);
+    DO_TEST("cpu-fallback", QEMU_CAPS_KVM);
+    DO_TEST_FAILURE("cpu-nofallback", QEMU_CAPS_KVM);
+    DO_TEST("cpu-strict1", QEMU_CAPS_KVM);
     DO_TEST("cpu-numa1", NONE);
     DO_TEST("cpu-numa2", QEMU_CAPS_SMP_TOPOLOGY);
     DO_TEST_PARSE_ERROR("cpu-numa3", NONE);
@@ -1284,7 +1284,10 @@ mymain(void)
     DO_TEST("pseries-usb-kbd", QEMU_CAPS_PCI_OHCI,
             QEMU_CAPS_DEVICE_USB_KBD, QEMU_CAPS_CHARDEV,
             QEMU_CAPS_DEVICE, QEMU_CAPS_NODEFCONFIG);
-    DO_TEST_FAILURE("pseries-cpu-exact", QEMU_CAPS_CHARDEV, QEMU_CAPS_DEVICE, QEMU_CAPS_NODEFCONFIG);
+    DO_TEST("pseries-cpu-exact", QEMU_CAPS_CHARDEV, QEMU_CAPS_DEVICE,
+            QEMU_CAPS_NODEFCONFIG);
+    DO_TEST("pseries-cpu-compat", QEMU_CAPS_KVM, QEMU_CAPS_CPU_HOST,
+            QEMU_CAPS_CHARDEV, QEMU_CAPS_DEVICE, QEMU_CAPS_NODEFCONFIG);
     DO_TEST("disk-ide-drive-split",
             QEMU_CAPS_DRIVE, QEMU_CAPS_DEVICE, QEMU_CAPS_NODEFCONFIG,
             QEMU_CAPS_IDE_CD);
diff -Naurp libvirt-1.2.8.orig/tests/qemuxml2xmloutdata/qemuxml2xmlout-graphics-spice-timeout.xml libvirt-1.2.8/tests/qemuxml2xmloutdata/qemuxml2xmlout-graphics-spice-timeout.xml
--- libvirt-1.2.8.orig/tests/qemuxml2xmloutdata/qemuxml2xmlout-graphics-spice-timeout.xml	2014-02-21 06:24:40.000000000 -0500
+++ libvirt-1.2.8/tests/qemuxml2xmloutdata/qemuxml2xmlout-graphics-spice-timeout.xml	2014-11-27 08:54:12.653689973 -0500
@@ -1,4 +1,4 @@
-<domain type='qemu'>
+<domain type='kvm'>
   <name>f14</name>
   <uuid>553effab-b5e1-2d80-dfe3-da4344826c43</uuid>
   <memory unit='KiB'>1048576</memory>
@@ -38,7 +38,7 @@
   <on_reboot>restart</on_reboot>
   <on_crash>restart</on_crash>
   <devices>
-    <emulator>/usr/bin/qemu</emulator>
+    <emulator>/usr/bin/qemu-kvm</emulator>
     <disk type='file' device='disk'>
       <driver name='qemu' type='qcow2'/>
       <source file='/var/lib/libvirt/images/f14.img'/>
